File: python/twelve-days/twelve_days.py
poems = [
    ("first", "a Partridge in a Pear Tree."),
    ("second", "two Turtle Doves, "),
    ("third", "three French Hens, "),
    ("fourth", "four Calling Birds, "),
    ("fifth", "five Gold Rings, "),
    ("sixth", "six Geese-a-Laying, "),
    ("seventh", "seven Swans-a-Swimming, "),
    ("eighth", "eight Maids-a-Milking, "),
    ("ninth", "nine Ladies Dancing, "),
    ("tenth", "ten Lords-a-Leaping, "),
    ("eleventh", "eleven Pipers Piping, "),
    ("twelfth", "twelve Drummers Drumming, ")
]

def recite(start_verse, end_verse):
    return [get_verse(num) for num in range(start_verse, end_verse + 1)]

def get_verse(num):
    verses = []
    if num == 1:
        verses = poems[0][1]
    else:
        verses = [poem[1] for poem in poems[num-1:0:-1]]
        verses.append(f"and {poems[0][1]}")

    opening = f"On the {poems[num - 1][0]} day of Christmas my true love gave to me: "
    return opening + ''.join(verses)

print(recite(12, 12))

File: go/twelve-days/twelve_days.go
// Package twelve is solution for problem Twelve Days.
package twelve

import (
	"fmt"
	"strings"
)

// Lyric represents a lyric part in twelve days song.
type Lyric struct {
	day  string
	text string
}

var lyrics = []Lyric{
	{
		day:  "first",
		text: "a Partridge in a Pear Tree.",
	},
	{
		day:  "second",
		text: "two Turtle Doves, and ",
	},
	{
		day:  "third",
		text: "three French Hens, ",
	},
	{
		day:  "fourth",
		text: "four Calling Birds, ",
	},
	{
		day:  "fifth",
		text: "five Gold Rings, ",
	},
	{
		day:  "sixth",
		text: "six Geese-a-Laying, ",
	},
	{
		day:  "seventh",
		text: "seven Swans-a-Swimming, ",
	},
	{
		day:  "eighth",
		text: "eight Maids-a-Milking, ",
	},
	{
		day:  "ninth",
		text: "nine Ladies Dancing, ",
	},
	{
		day:  "tenth",
		text: "ten Lords-a-Leaping, ",
	},
	{
		day:  "eleventh",
		text: "eleven Pipers Piping, ",
	},
	{
		day:  "twelfth",
		text: "twelve Drummers Drumming, ",
	},
}

// Song returns a full twelve days song.
func Song() string {
	song := make([]string, 0)
	for i := range lyrics {
		song = append(song, Verse(i+1))
	}
	return strings.Join(song, "\n")
}

// Verse returns from 1 to n-th verses of the twelve days song.
func Verse(n int) string {
	var result strings.Builder
	begin := fmt.Sprintf("On the %s day of Christmas my true love gave to me: ", lyrics[n-1].day)
	result.WriteString(begin)
	for i := n - 1; i >= 0; i-- {
		s := fmt.Sprintf("%s", lyrics[i].text)
		result.WriteString(s)
	}
	return result.String()
}
